Hibernate reactive panache refactoring

mkouba opened the pull request:
- [x] do not store the current reactive session in the CDI request context but instead in the vertx duplicated context
- [x] do not offload execution of a panache entity method on the current vertx context but instead validate that the method is executed on the vetx duplicated context
- [x] introduce `@WithSession`, `@WithSessionOnDemand` and `@WithTransaction` bindings and interceptors
- [x] deprecate `ReactiveTransactional` (`WithTransaction` should be used instead)
- [x] `ReactiveTransactionalInterceptor` can only be used for methods that return `Uni`; this is validated at build time
- [x] if `resteasy-reactive` is present then automatically add the `@WithSessionOnDemand` binding to all resource methods on classes that use a panache entity
- [x] remove the quarkus-integration-test-hibernate-reactive-panache-blocking module
- [x] `quarkus-test-vertx` - run the test method on a duplicated vertx context even if the `@RunOnVertxContext` is not present but the `@TestReactiveTransaction` is
- [x] update `integration-tests/hibernate-reactive-panache-kotlin`
- [x] update documentation

mkouba committed:
Hibernate reactive panache refactoring

- do not store the current reactive session in the CDI request context
but instead in the vertx duplicated context
- do not offload execution of a panache entity method on the current
vertx context but instead validate that the method is executed on the
vetx duplicated context
- introduce WithSession, WithSessionOnDemand and WithTransaction
bindings and interceptors
- deprecate ReactiveTransactional
- ReactiveTransactionalInterceptor can only be used for methods that
return Uni/Multi; this is validated at build time
- if resteasy-reactive is present then automatically add WithSessionOnDemand binding to all resource methods on classes that use a panache entity
- also remove the quarkus-integration-test-hibernate-reactive-panache-blocking module
- quarkus-test-vertx - run the test method on a duplicated vertx context even if the RunOnVertxContext is not present but the TestReactiveTransaction is

mkouba commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/ReactiveSession.java:
```diff
@@ -0,0 +1,30 @@
+@InterceptorBinding
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(value = RetentionPolicy.RUNTIME)
+public @interface ReactiveSession {
```
I think that `@WithinReactiveSession` would be a more meaningful name... although it's quite long. OTOH it should not be necessary in most cases.

mkouba replied:
We now have `@WithSession` (opens a new session or reuses existing) and `@WithSessionOnDemand` (opens a session or reuses existing _if needed_).

mkouba commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -0,0 +1,194 @@
+                // the context was marked as "lazy" which means that the session will be eventually closed
+                return getSessionFactory().openSession().invoke(s -> context.putLocal(key, s));
+            } else {
+                throw new IllegalStateException("No current Mutiny.Session found");
```
We should use a more actionable message here...

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -0,0 +1,214 @@
+            return s.withTransaction(t -> work.get()
+                    .onFailure()
+                    // close the session and remove it from the vertx duplicated context
+                    .invoke(e -> closeSession()));
```
Why do we have to close the session here? Doesn't `withSession` do it?

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -0,0 +1,214 @@
+            return s.withTransaction(t -> work.apply(t)
+                    .onFailure()
+                    // close the session and remove it from the vertx duplicated context
+                    .invoke(e -> closeSession()));
```
Same question here.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -0,0 +1,214 @@
+            // reactive session does not exist - open a new one and close it when the returned Uni completes
+            return getSessionFactory()
+                    .openSession()
+                    .invoke(s -> context.putLocal(key, s))
```
Doesn't the `Mutiny.SessionFactory` already associate it to the context?

mkouba replied:
`openSession()` itself does not store the session in the context. Only `SessionFactory.withSession()` does.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -0,0 +1,214 @@
+            if (context.getLocal(LAZY_SESSION_KEY) != null) {
+                // open a new reactive session and store it in the vertx duplicated context
+                // the context was marked as "lazy" which means that the session will be eventually closed
+                return getSessionFactory().openSession().invoke(s -> context.putLocal(key, s));
```
Same question here.

FroMage commented on integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java:
```diff
@@ -234,46 +236,67 @@ public void testTestTransaction2(UniAsserter asserter) {
+        asserter.assertEquals(() -> testReactiveTransactional3(), 1l);
+    }
+
+    @ReactiveTransactional
```
So `@ReactiveTransactional` can no longer be used on test methods?

mkouba replied:
It can only be used if the method returns `Uni`. However, the `TestReactiveTransaction` works as before.

FroMage commented on integration-tests/pom.xml:
```diff
@@ -190,7 +190,6 @@
                 <module>hibernate-reactive-postgresql</module>
                 <module>hibernate-reactive-panache</module>
                 <module>hibernate-reactive-panache-kotlin</module>
-                <module>hibernate-reactive-panache-blocking</module>
```
What happened to this one?

mkouba replied:
This module was testing the functionality that got removed, see https://github.com/quarkusio/quarkus/blob/2.15.0.Final/integration-tests/hibernate-reactive-panache-blocking/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java#L11-L16.

FroMage commented on integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/TestReactiveTransactionTest.java:
```diff
@@ -6,17 +6,20 @@
 public class TestReactiveTransactionTest {
 
+    @RunOnVertxContext
     @TestReactiveTransaction
```
Can we make that one imply the previous one?

mkouba replied:
Hmm, `@RunOnVertxContext` is not an interceptor binding so unfortunately we can't leverage the fact that interceptor bindings are transitive. And the `io.quarkus.test.vertx.RunOnVertxContextTestMethodInvoker` is using reflection to detect this annotation so we can't transform the annotations either.

But it seems that `RunOnVertxContextTestMethodInvoker` is already used for methods annotated with `TestReactiveTransaction` so the only thing we're missing is the [vertx duplicated context](https://github.com/quarkusio/quarkus/blob/2.15.0.Final/test-framework/vertx/src/main/java/io/quarkus/test/vertx/RunOnVertxContextTestMethodInvoker.java#L99-L109). 

@geoand Do you think it would make sense to return `true` on [this line](https://github.com/quarkusio/quarkus/blob/2.15.0.Final/test-framework/vertx/src/main/java/io/quarkus/test/vertx/RunOnVertxContextTestMethodInvoker.java#LL105C11-L105C11)? I.e. use the duplicated context by default?

mkouba replied:
Actually, we should only return `true` if `@RunOnVertxContext` is not present, but `@TestReactiveTransaction` is...

geoand replied:
I think that makes sense

mkouba replied:
Ok, I'll try to update the code and we'll see if the change makes it in ;-)

FroMage commented:
Sorry for the late review :(

mkouba commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/ReactiveTransactional.java:
```diff
@@ -11,16 +11,13 @@
- *
- * Otherwise, invocations are only allowed when not running from a {@link VertxThread} and the behaviour
- * will be the same as a blocking call to {@link Mutiny.Session#withTransaction(java.util.function.Function)}.
+ * TODO this annotation should go into io.quarkus.hibernate.reactive.panache.common
```
For now the `ReactiveTransactional` is deprecated (but still works) and users are encouraged to use `@WithTransaction` instead...

mkouba commented:
> Sorry for the late review :(

No problem, better late than never! ;-)

FroMage reviewed (approved):
LGTM.
Do you think there's any need for a `@WithoutSession` to disable the automatic `@WithSessionIfNeeded`? I don't think so.
In Renarde, I automatically add the `@Transactional` annotation to `POST/PUT/DELETE` methods, but I've often wondered if I should add a `@NoTransaction` annotation to disable this. Perhaps we need a `@WithoutTransaction` in HR/Panache, so that I can avoid auto-adding `@WithTransaction` when I integrate HR in Renarde?

mkouba commented:
> Do you think there's any need for a @WithoutSession to disable the automatic @WithSessionIfNeeded? I don't think so.

I agree.

> In Renarde, I automatically add the `@Transactional` annotation to `POST/PUT/DELETE` methods, but I've often wondered if I should add a `@NoTransaction` annotation to disable this. Perhaps we need a `@WithoutTransaction` in HR/Panache, so that I can avoid auto-adding `@WithTransaction` when I integrate HR in Renarde?

What would be the point of `@WithoutTransaction` in the context of HR/Panache? I mean this is the default, unless you specify `@WithTransaction` so I don't really understand...

mkouba commented:
Well, I think that such an annotation should go into Renard then...

mkouba commented:
> If it is indeed the case (confirmed) then may be it can make sense to have `quarkus-hibernate-panache-reactive` directly depending on `quarkus-reactive-pg-client` ? In devmode, it would be nice to be able to only add `quarkus-hibernate-panache-reactive` to the pom, without having to add `quarkus-reactive-pg-client`, if it is required anyway...

I don't think it's required. I mean you just need to add an [reactive SQL client extension](https://quarkus.io/guides/reactive-sql-clients) like for any [Hibernate Reactive](https://quarkus.io/guides/hibernate-reactive#hr-getting-started) application.

sberyozkin commented:
But if without it NPE is returned then may be it is required ?

sberyozkin commented:
Sorry, see what you mean, there are a few client types, may be a better error message can be returned like please add a client, etc

mkouba commented:
> Sorry, see what you mean, there are a few client types, may be a better error message can be returned like please add a client, etc

I wonder if the NPE is thrown for vanilla Hibernate Reactive as well?

DavideD commented:
Is there a stack trace with the NullPointer?

DavideD commented:
In Hibernate Reactive, if the vertx client library is missing you should see a message like this:
```
Caused by: org.hibernate.internal.util.config.ConfigurationException: No suitable drivers found for URI scheme: postgresql
```

mkouba commented:
> In Hibernate Reactive, if the vertx client library is missing you should see a message like this:
> 
> ```
> Caused by: org.hibernate.internal.util.config.ConfigurationException: No suitable drivers found for URI scheme: postgresql
> ```

If I configure the datasource and forget to include the `quarkus-reactive-pg-client` then I get `Failed to start application (with profile prod): java.lang.IllegalStateException: No pool has been defined for persistence unit default-reactive`.

Anyway, could we move the discussion to a separate issue? It's not related to this pull request.

mkouba commented:
There is one more open question: what do we do with `PanacheEntityBase.stream()` methods.

I've opened a discussion on zulip: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Hibernate.20reactive.20panache.20and.20Multi

I'm still not sure what to do but I like Julien's comment:

> Julien Ponge: Uni<List<Fruits>> has one big advantage: we aren't dressing what is not a stream into a stream :-)
>
> Julien Ponge: Multi shall really be exposed for back-pressured streams, not for something that's already in memory
>
> Julien Ponge: Java Streams (not RS/Flow) works great for anything in-memory. It doesn't have the overhead of a backpressured protocol.
>
>Julien Ponge: Now you can always wrap an in-memory stream as a Multi and do what other vendors do by wrapping reactive streams on top of anything asynchronous, but I think it's worth being precise on what gets exposed in APIs. Uni = one-shot async I/O operation, Multi = back-pressured stream (possibly infinite) of I/O events, JDK Streams = in-memory stream (hence finite)

mkouba commented:
> The options are:
> 
>     1. Remove the methods completely
> 
>     2. Add `Panache.withSession()` variants for `Multi` and add support for `Multi` to the `@WithSession` interceptor.

This option is now implemented. We have two methods `Panache.uniWithSession()` and `Panache.multiWithSession()`.

gsmet commented:
@Sanne @DavideD is this good for you?

Wondering if we should make it 3.0 material.

mkouba commented:
> @Sanne @DavideD is this good for you?
> 
> Wondering if we should make it 3.0 material.

That is the plan ;-)

This PR is ready for final review. It's a large changeset and there might be inconsistencies here and there but I believe that it's good enough to get it in and get some feedback in early phases of 3.0.

mkouba commented:
I've just converted this PR back to the draft after [this zulip discussion with @Sanne](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Hibernate.20reactive.20panache.20and.20Multi/near/324794621).

It seems that users are encouraged to use a transaction for anything more complex than a single `select` query. See also [Why always having a transaction manager?](https://quarkus.io/guides/transaction#why-always-having-a-transaction-manager).

Given the fact that we can't provide something like `Panache.multiWithTransaction(Supplier<Multi<T>> multiSupplier)` (because HR only supports `Mutiny.Session.withTransaction(Function<Transaction, Uni<T>>)`) it does not make much sense to keep the `PanacheEntityBase.stream()` methods IMO.

CC @FroMage @DavideD @Sanne

FroMage commented:
> Given the fact that we can't provide something like Panache.multiWithTransaction(Supplier<Multi<T>> multiSupplier) (because HR only supports Mutiny.Session.withTransaction(Function<Transaction, Uni<T>>)) it does not make much sense to keep the PanacheEntityBase.stream() methods IMO.

Well, we can always implement it ourselves, it's not like it's a complicated method.

FroMage commented:
But again, I'm not opposed at all to dropping `Multi` if everyone wants that ;)

mkouba commented:
> > Given the fact that we can't provide something like Panache.multiWithTransaction(Supplier<Multi> multiSupplier) (because HR only supports Mutiny.Session.withTransaction(Function<Transaction, Uni>)) it does not make much sense to keep the PanacheEntityBase.stream() methods IMO.
> 
> Well, we can always implement it ourselves, it's not like it's a complicated method.

Well, you would need to work with the `io.vertx.sqlclient.SqlConnection` directly and bypass the HR API. Which is IMO :-1: .

> But again, I'm not opposed at all to dropping Multi if everyone wants that ;)

That's the problem. It does not seem that _everyone_ wants to do it but nobody suggests a reasonable solution :D

mkouba committed:
Hibernate reactive panache - remove all Multi<T> stream() methods

- neither Hibernate Reactive nor reactive clients support streaming
- furthermore, we're not able to provide a Panache#withTransaction()
alternative for Multi without bypassing Hibernate Reactive API

mkouba commented:
The CI is green now!

FTR all `Multi<T> stream()` methods were removed.

@FroMage @DavideD @Sanne I'm going to merge this pull reuqest but I do expect some follow-ups based on potential feedback.

DavideD commented:
Thanks @mkouba

FroMage commented:
+1
